### PR TITLE
Fix race condition while disposing actor and avoid unecessary allocation

### PR DESCRIPTION
**Pull request by youngbupark**
# Description

* Fix race condition - https://github.com/dapr/dapr/issues/2306
* Do not process drained actor
* Avoid unnecessary channel allocation
* Rename member name to be explicit and add more comments
* Add more tests

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #2306

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/actors/actor_test.go [hunk 1]
```diff
@@ -33,39 +33,109 @@ func TestConcurrencyLocks(t *testing.T) {
...
 	assert.Equal(t, firstLockTime, testActor.lastUsedTime)
 
 	// unlock the first lock
-	testActor.unLock()
+	testActor.unlock()
 
 	time.Sleep(10 * time.Millisecond)
+	assert.Equal(t, int32(1), testActor.pendingActorCalls)
 	assert.True(t, testActor.isBusy())
 	assert.NotEqual(t, firstLockTime, testActor.lastUsedTime)
```
Comment (mukundansundar) [inline pkg/actors/actor_test.go:46]: check `assert.True(t, testActor.After(firstLockTime.Add(10*time.Millisecond)))`

File: pkg/actors/actors_test.go [hunk 2]
```diff
@@ -700,6 +702,41 @@ func TestDeleteState(t *testing.T) {
...
+	t.Run("actor is already disposed", func(t *testing.T) {
+		testActorRuntime := newTestActorsRuntime()
+		actorKey := testActorRuntime.constructCompositeKey(testActorType, testActorID)
+		act := newActor(testActorType, testActorID)
+		testActorRuntime.actorsTable.LoadOrStore(actorKey, act)
+		act.lock()
+		assert.True(t, act.isBusy())
+		ch := act.channel()
+		act.unlock()
+		_, closed := <-ch
```
Comment (mukundansundar) [inline pkg/actors/actors_test.go:730]: rename closed to more. Preferably add comment on what the second return value is.

File: pkg/actors/actor.go [hunk 3]
```diff
@@ -11,56 +11,87 @@ import (
...
+	// actorType is the type of actor.
 	actorType string
-	actorID   string
+	// actorID is the ID of actorType.
+	actorID string
+
+	// concurrencyLock is the lock to maintain actor's turn-based concurrency.
+	concurrencyLock *sync.Mutex
+	// pendingActorCalls is the number of the current pending actor calls by turn-based concurrency.
+	pendingActorCalls int32
```
Comment (tcnghia) [inline pkg/actors/actor.go:32]: If you use https://godoc.org/go.uber.org/atomic#Int32 we are forced to use atomic operations by default and avoid future typos

File: pkg/actors/actor.go [hunk 4]
```diff
@@ -11,56 +11,87 @@ import (
...
 }
 
-func (a *actor) unLock() {
-	if a.busy {
-		a.busy = false
-		close(a.busyCh)
+// unlock release the lock for turn-based concurrency. If disposeCh is available,
+// it will close the channel to notify runtime to dispose actor.
+func (a *actor) unlock() {
+	if atomic.AddInt32(&a.pendingActorCalls, int32(-1)) == 0 {
```
Comment (tcnghia) [inline pkg/actors/actor.go:89]:
```suggestion
	if atomic.AddInt32(&a.pendingActorCalls, -1) == 0 {
```
I think consts are untyped by default so this should work. Same with other places.

File: pkg/actors/actor.go [hunk 5]
```diff
@@ -11,56 +11,87 @@ import (
...
+// lock holds the lock for turn-based concurrency.
+func (a *actor) lock() error {
+	atomic.AddInt32(&a.pendingActorCalls, int32(1))
+	diag.DefaultMonitoring.ReportActorPendingCalls(a.actorType, a.pendingActorCalls)
 	a.concurrencyLock.Lock()
-
-	a.busy = true
-	a.busyCh = make(chan bool, 1)
+	if a.disposed {
+		return ErrActorDisposed
```
Comment (tcnghia) [inline pkg/actors/actor.go:80]: In case of error, do we still want to increase (line 76)?

Comment (artursouza) [reply pkg/actors/actor.go:80]: Line 76 seems to be decreased on unlock.

**Review by artursouza — DISMISSED** [review]

File: pkg/actors/actor.go (hunk 5 above)
Comment (youngbupark) [reply pkg/actors/actor.go:80]: unlock needs to be called regardless of error, but I think calling unlock before return error would be safer.

File: pkg/actors/actor_test.go (hunk 1 above)
Comment (youngbupark) [reply pkg/actors/actor_test.go]: After returns channel :) anyhow, I change to check the duration.

File: pkg/actors/actor.go (hunk 4 above)
Comment (youngbupark) [reply pkg/actors/actor.go]: I used atomic wrapper Inc/Dec.

File: pkg/actors/actor.go (hunk 5 above)
Comment (youngbupark) [reply pkg/actors/actor.go:82]: fixed and added test

File: pkg/actors/actor.go (hunk 3 above)
Comment (youngbupark) [reply pkg/actors/actor.go]: switched to atomic wrapper. thanks.

File: pkg/actors/actors_test.go (hunk 2 above)
Comment (youngbupark) [reply pkg/actors/actors_test.go:730]: I would not add comment for golang channel syntax... closed returns channel status.

**Review by tcnghia — APPROVED** [review]

**Pull request merged**